src/compiler/checker.ts

```diff
@@ -14001,10 +13993,10 @@ namespace ts {
                 const id = objectType.id + "," + indexType.id;
                 let type = indexedAccessTypes.get(id);
                 if (!type) {
-                    indexedAccessTypes.set(id, type = createIndexedAccessType(objectType, indexType, aliasSymbol, aliasTypeArguments));
+                    indexedAccessTypes.set(id, type = createIndexedAccessType(objectType, indexType, aliasSymbol, aliasTypeArguments, accessFlags));
                 }
 
-                return shouldIncludeUndefined ? getUnionType([type, undefinedType]) : type;
+                return type;

```

Comment:
At this point, we don’t yet know whether the index type will be constrained to something that resolves to property signatures or index signatures, so we can’t return a union with undefined. But when we come back to do the deferred checking of the indexed access type we saved, we’ve lost the `accessFlags` that indicate that this is a position where `undefined` should be included in the type.

As I’m writing this, I’m realizing that this is probably broken for generic indexed access types that get used both for reading and for writing, so converting back to draft.

Reply:
The only easy way I could find to fix this was creating separate IndexedAccessTypes (and keying the cache) depending on whether it was an expression where `--noUncheckedIndexedAccess` could have an effect or not. So in effect, in

```ts
function<T extends string>(key: T) {
  obj[key] = undefined;
  return obj[key];
}
```

the two usages of `obj[key]` have two separate types. (This only happens when the object type and/or the index type is generic.) I think this is technically possible to avoid, but would involve a _lot_ of inexpedient threading of `accessFlags` (or some parameter based on it) through a bunch of functions.